<!DOCTYPE html>
<html>
<head>
	<meta charset='utf-8'>
	<title>Bioconductor RNA-Seq workflow</title>
	<meta name="description" content="Bioconductor RNA-Seq workflow">
	<meta name="author" content="Michael Love">
	<meta name="generator" content="slidify" />
	
	<!-- LOAD STYLE SHEETS -->
	<link rel="stylesheet" href="libraries/frameworks/html5slides/default/styles.css">
	<link rel="stylesheet" href="libraries/highlighters/highlight.js/css/tomorrow.css" />  <link rel="stylesheet" href = "assets/css/ribbons.css">
  
</head>
<body style='display: none'>
	<section class='slides layout-regular template-regular'>
     <article class="" id="slide-1" style="background:;">
  <h3>Bioconductor RNA-Seq workflow</h3>
  <ol>
<li>preparing gene models</li>
<li>read counting</li>
<li>EDA (exploratory data analysis)</li>
<li>differential expression analysis</li>
<li>annotating results</li>
</ol>

</article>

<article class="" id="slide-2" style="background:;">
  <h3>Preparing gene models part 1</h3>
  <p><a href="http://www.bioconductor.org/packages/release/bioc/html/GenomicFeatures.html">GenomicFeatures</a></p>

<pre><code class="r">library( &quot;GenomicFeatures&quot; )
# takes ~10 min
txdb &lt;- makeTranscriptDbFromBiomart(
  biomart=&quot;ensembl&quot;,
  dataset=&quot;hsapiens_gene_ensembl&quot; )
</code></pre>

<p>smart to use <code>saveDb()</code> to only do this once</p>

</article>

<article class="" id="slide-3" style="background:;">
  <h3>Preparing gene models part 2</h3>
  <ul>
<li><code>makeTranscriptDbFromGFF()</code> accepts GTF</li>
<li><code>library(TxDb.Hsapiens.UCSC.hg19.knownGene)</code> ready to go</li>
<li>soon also, <code>AnnotationDbi</code> will offer ready to go</li>
</ul>

</article>

<article class="" id="slide-4" style="background:;">
  <h3>Super useful</h3>
  <pre><code class="r">seqlevelsStyle(gr) &lt;- &quot;NCBI&quot;
seqlevelsStyle(gr) &lt;- &quot;UCSC&quot;
</code></pre>

</article>

<article class="" id="slide-5" style="background:;">
  <h3>What&#39;s in txdb?</h3>
  <p><code>sqlite3 txdb.sqlite &#39;.tables&#39;</code></p>

<p><code>cds exon metadata transcript chrominfo gene splicing</code></p>

</article>

<article class="" id="slide-6" style="background:;">
  <h3>What&#39;s in txdb?</h3>
  <pre><code class="r">library(TxDb.Hsapiens.UCSC.hg19.knownGene)
TxDb.Hsapiens.UCSC.hg19.knownGene
</code></pre>

<pre><code>## TranscriptDb object:
## | Db type: TranscriptDb
## | Supporting package: GenomicFeatures
## | Data source: UCSC
## | Genome: hg19
## | Organism: Homo sapiens
## | UCSC Table: knownGene
## | Resource URL: http://genome.ucsc.edu/
## | Type of Gene ID: Entrez Gene ID
## | Full dataset: yes
## | miRBase build ID: GRCh37
## | transcript_nrow: 82960
## | exon_nrow: 289969
## | cds_nrow: 237533
## | Db created by: GenomicFeatures package from Bioconductor
## | Creation time: 2014-03-17 16:15:59 -0700 (Mon, 17 Mar 2014)
## | GenomicFeatures version at creation time: 1.15.11
## | RSQLite version at creation time: 0.11.4
## | DBSCHEMAVERSION: 1.0
</code></pre>

</article>

<article class="" id="slide-7" style="background:;">
  <h3>Extract exons for each gene</h3>
  <pre><code class="r"># takes ~30 seconds
exonsByGene &lt;- exonsBy( txdb, by=&quot;gene&quot; )
</code></pre>

</article>

<article class="" id="slide-8" style="background:;">
  <h3>Extract exons for each gene</h3>
  <pre><code class="r"># TODO show exonsByGene
</code></pre>

</article>

<article class="" id="slide-9" style="background:;">
  <h3>Read counting</h3>
  <pre><code class="r">library( &quot;Rsamtools&quot; )
bamLst &lt;- BamFileList( fls, yieldSize=2000000 )
</code></pre>

</article>

<article class="" id="slide-10" style="background:;">
  <h3>Read counting</h3>
  <pre><code class="r">library( &quot;GenomicAlignments&quot; )
register( MulticoreParam( workers=4 ) )
se &lt;- summarizeOverlaps( features=exonsByGene,
                        reads=bamLst,
                        mode=&quot;Union&quot;,
                        singleEnd=FALSE,
                        ignore.strand=TRUE,
                        fragments=TRUE )
</code></pre>

</article>

<article class="" id="slide-11" style="background:;">
  <h3>SummarizedExperiment</h3>
  <p><center><img src="summarizedexperiment.png" width=700/></center></p>

</article>

<article class="" id="slide-12" style="background:;">
  <h3>Metadata stored without user effort</h3>
  <pre><code class="r">metadata( rowData( se ) )
</code></pre>

</article>

<article class="" id="slide-13" style="background:;">
  <h3>Exploratory data analysis (EDA)</h3>
  
</article>

<article class="" id="slide-14" style="background:;">
  <h3>Differential expression analysis</h3>
  
</article>

<article class="" id="slide-15" style="background:;">
  <h3>The generalized linear model part 1</h3>
  <p>\[ K_{ij} \sim \text{NB}( \mu_{ij}, \alpha_i )  \]</p>

<ul>
<li>Read count \(K_{ij}\) for gene <em>i</em> sample <em>j</em>.</li>
<li>2 parameter count distribution: mean \(\mu\), dispersion \(\alpha\)</li>
</ul>

<p>\[ \mu_{ij} = s_{ij} q_{ij} \]</p>

<ul>
<li>Normalized by size factor \(s_{ij}\), with \(q_{ij}\) remaining</li>
<li>Often size factor \(s_j\)</li>
</ul>

</article>

<article class="" id="slide-16" style="background:;">
  <h3>The generalized linear model part 2</h3>
  <p>\[ \log q_{ij} = \sum_r x_{jr} \beta_{ir} \]</p>

<p>\[ \begin{array}{c}
\log q_1 \\
\log q_2 \\
\log q_3 \\
\log q_4 \end{array}
= \left( \begin{array}{cc}
1 & 0 \\
1 & 0 \\
1 & 1 \\
1 & 1 \end{array} \right)
\left( \begin{array}{c}
\beta_0 \\
\beta_1 \end{array} \right) \] </p>

<ul>
<li>results are logarithm base 2 fold changes </li>
</ul>

</article>

<article class="" id="slide-17" style="background:;">
  <h3>Multigroup comparisons in DESeq2</h3>
  <p>\[ \begin{array}{c}
\log q_1 \\
\log q_2 \\
\log q_3 \\
\log q_4 \\
\log q_5 \\
\log q_6 \end{array}
= \left( \begin{array}{cccc}
1 & 1 & 0 & 0 \\
1 & 1 & 0 & 0 \\
1 & 0 & 1 & 0 \\
1 & 0 & 1 & 0 \\
1 & 0 & 0 & 1 \\
1 & 0 & 0 & 1 \\ \end{array} \right)
\left( \begin{array}{c}
\beta_0 \\
\beta_1 \\
\beta_2 \\
\beta_3 \end{array} \right) \] </p>

</article>

<article class="" id="slide-18" style="background:;">
  <h3>Normalization for sample-specific GC and transcript length</h3>
  <p><center><img src="cqn.png" width=700/></center></p>

</article>

<article class="" id="slide-19" style="background:;">
  <h3>Controlling for unknown batch</h3>
  <p><center><img src="sva.png" width=700/></center></p>

<p><a href="http://dx.doi.org/10.1371/journal.pgen.0030161">Leek and Storey (2007)</a></p>

</article>

<article class="" id="slide-20" style="background:;">
  <h3>ReportingTools</h3>
  <p><center><img src="reptools.png" width=700/></center></p>

</article>

<article class="" id="slide-21" style="background:;">
  <h3>Support</h3>
  <ul>
<li>mailing list:

<ol>
<li>package name in title</li>
<li>describe experiment, what&#39;s the <em>biological question</em></li>
<li>provide code</li>
<li>sessionInfo()</li>
</ol></li>
<li><code>browseVignettes(&quot;pkg&quot;)</code></li>
<li><code>?function</code></li>
</ul>

</article>

<article class="" id="slide-22" style="background:;">
  <h3>Thanks</h3>
  <ul>
<li>Bioconductor core team</li>
<li><em>DESeq</em>/<em>DEXSeq</em> team

<ul>
<li>Simon Anders</li>
<li>Alejandro Reyes</li>
<li>Wolfgang Huber</li>
</ul></li>
<li>Rafael Irizarry</li>
</ul>

</article>

  </section>
</body>
  <!-- LOAD JAVASCRIPTS  -->
	<script src='libraries/frameworks/html5slides/default/slides.js'></script>
	<!-- MathJax: Fall back to local if CDN offline but local image fonts are not supported (saves >100MB) -->
	<script type="text/x-mathjax-config">
	  MathJax.Hub.Config({
	    tex2jax: {
	      inlineMath: [['$','$'], ['\\(','\\)']],
	      processEscapes: true
	    }
	  });
	</script>
	<script type="text/javascript" src="http://cdn.mathjax.org/mathjax/2.0-latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
	<!-- <script src="https://c328740.ssl.cf1.rackcdn.com/mathjax/2.0-latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
	</script> -->
	<script>window.MathJax || document.write('<script type="text/x-mathjax-config">MathJax.Hub.Config({"HTML-CSS":{imageFont:null}});<\/script><script src="libraries/widgets/mathjax/MathJax.js?config=TeX-AMS-MML_HTMLorMML"><\/script>')
</script>
<!-- LOAD HIGHLIGHTER JS FILES -->
	<script src="libraries/highlighters/highlight.js/highlight.pack.js"></script>
	<script>hljs.initHighlightingOnLoad();</script>
	<!-- DONE LOADING HIGHLIGHTER JS FILES -->
	 
	</html>
