<!DOCTYPE html>
<html>
<head>
	<meta charset='utf-8'>
	<title>Bioconductor RNA-Seq workflow</title>
	<meta name="description" content="Bioconductor RNA-Seq workflow">
	<meta name="author" content="Michael Love">
	<meta name="generator" content="slidify" />
	
	<!-- LOAD STYLE SHEETS -->
	<link rel="stylesheet" href="libraries/frameworks/html5slides/default/styles.css">
	<link rel="stylesheet" href="libraries/highlighters/highlight.js/css/tomorrow.css" />  <link rel="stylesheet" href = "assets/css/ribbons.css">
  
</head>
<body style='display: none'>
	<section class='slides layout-regular template-regular'>
     <article class="" id="slide-1" style="background:;">
  <h3>Bioconductor RNA-Seq workflow</h3>
  <ol>
<li>preparing gene models</li>
<li>read counting</li>
<li>EDA (exploratory data analysis)</li>
<li>differential expression analysis</li>
<li>annotating results</li>
</ol>

</article>

<article class="" id="slide-2" style="background:;">
  <h3>Preparing gene models</h3>
  <pre><code class="r">library( &quot;GenomicFeatures&quot; )
# takes ~10 min
txdb &lt;- makeTranscriptDbFromBiomart( biomart=&quot;ensembl&quot;,
                                    dataset=&quot;hsapiens_gene_ensembl&quot; )
</code></pre>

<p>smart to use <code>saveDb()</code> to only do this once</p>

<ul>
<li><a href="http://www.bioconductor.org/packages/release/bioc/html/GenomicFeatures.html">GenomicFeatures</a></li>
</ul>

</article>

<article class="" id="slide-3" style="background:;">
  <h3>Preparing gene models</h3>
  <ul>
<li><code>makeTranscriptDbFromGFF()</code> accepts GTF</li>
<li><code>library(TxDb.Hsapiens.UCSC.hg19.knownGene)</code> ready to go</li>
<li>soon also, <code>AnnotationHub</code> will offer ready to go</li>
</ul>

</article>

<article class="" id="slide-4" style="background:;">
  <h3>Super useful</h3>
  <pre><code class="r">seqlevelsStyle(gr) &lt;- &quot;NCBI&quot;
seqlevelsStyle(gr) &lt;- &quot;UCSC&quot;
</code></pre>

</article>

<article class="" id="slide-5" style="background:;">
  <h3>Extract exons for each gene</h3>
  <pre><code class="r"># takes ~30 seconds
exonsByGene &lt;- exonsBy( txdb, by=&quot;gene&quot; )
</code></pre>

<pre><code class="r">head( exonsByGene[[1]] )
</code></pre>

<pre><code>## GRanges with 6 ranges and 2 metadata columns:
##       seqnames               ranges strand |   exon_id       exon_name
##          &lt;Rle&gt;            &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;     &lt;character&gt;
##   [1]        X [99883667, 99884983]      - |    664095 ENSE00001459322
##   [2]        X [99885756, 99885863]      - |    664096 ENSE00000868868
##   [3]        X [99887482, 99887565]      - |    664097 ENSE00000401072
##   [4]        X [99887538, 99887565]      - |    664098 ENSE00001849132
##   [5]        X [99888402, 99888536]      - |    664099 ENSE00003554016
##   [6]        X [99888402, 99888536]      - |    664100 ENSE00003658801
##   ---
##   seqlengths:
##                    1                 2 ...            LRG_99
##            249250621         243199373 ...             13294
</code></pre>

</article>

<article class="" id="slide-6" style="background:;">
  <h3>Read counting</h3>
  <pre><code class="r">library( &quot;Rsamtools&quot; )
bamLst &lt;- BamFileList( fls, yieldSize=2000000 )
</code></pre>

<ul>
<li><a href="http://www.bioconductor.org/packages/release/bioc/html/Rsamtools.html">Rsamtools</a></li>
</ul>

</article>

<article class="" id="slide-7" style="background:;">
  <h3>Read counting</h3>
  <pre><code class="r">library( &quot;GenomicAlignments&quot; )
register( MulticoreParam( workers=4 ) )
# takes e.g. ~30 minutes per sample for 40 million PE reads 
se &lt;- summarizeOverlaps( features=exonsByGene,
                        reads=bamLst,
                        mode=&quot;Union&quot;,
                        singleEnd=FALSE,
                        ignore.strand=TRUE,
                        fragments=TRUE )
</code></pre>

<ul>
<li><a href="http://www.bioconductor.org/packages/release/bioc/html/GenomicAlignments.html">GenomicAlignments</a></li>
<li><a href="http://www-huber.embl.de/users/anders/HTSeq/doc/count.html">htseq-count</a>: python library</li>
<li><a href="http://www.bioconductor.org/packages/release/bioc/html/Rsubread.html">featureCounts</a>: <code>featureCounts()</code></li>
</ul>

</article>

<article class="" id="slide-8" style="background:;">
  <h3>SummarizedExperiment</h3>
  <p><center><img src="summarizedexperiment.png" width=700/></center></p>

</article>

<article class="" id="slide-9" style="background:;">
  <h3>Metadata stored without user effort</h3>
  <pre><code class="r">metadata( rowData( se ) )[[ 1 ]][ 1:6 ]
</code></pre>

<pre><code>## $`Db type`
## [1] &quot;TranscriptDb&quot;
## 
## $`Supporting package`
## [1] &quot;GenomicFeatures&quot;
## 
## $`Data source`
## [1] &quot;BioMart&quot;
## 
## $Organism
## [1] &quot;Homo sapiens&quot;
## 
## $`Resource URL`
## [1] &quot;www.biomart.org:80&quot;
## 
## $`BioMart database`
## [1] &quot;ensembl&quot;
</code></pre>

</article>

<article class="" id="slide-10" style="background:;">
  <h3>Add sample data</h3>
  <pre><code class="r">samples &lt;- read.csv( &quot;sample_data.csv&quot; )
colData( se ) &lt;- DataFrame( samples )
</code></pre>

</article>

<article class="" id="slide-11" style="background:;">
  <h3>Exploratory data analysis (EDA)</h3>
  <pre><code class="r">dds &lt;- DESeqDataSet( se, ~ group + condition )
rld &lt;- rlog( dds )
plotPCA( rld )
</code></pre>

<p><center><img src="pca.png" width=500/></center></p>

</article>

<article class="" id="slide-12" style="background:;">
  <h3>Differential expression analysis</h3>
  <ul>
<li><a href="http://www.bioconductor.org/packages/release/bioc/html/DESeq2.html">DESeq2</a></li>
<li><a href="http://www.bioconductor.org/packages/release/bioc/html/DEXSeq.html">DEXSeq</a>
differential exon usage</li>
<li><a href="http://www.bioconductor.org/packages/release/bioc/html/edgeR.html">edgeR</a></li>
<li><a href="http://www.bioconductor.org/packages/release/bioc/html/limma.html">limma</a> + voom normalization</li>
<li><a href="http://www.bioconductor.org/packages/release/bioc/html/DSS.html">DSS</a></li>
<li><a href="http://www.bioconductor.org/packages/release/bioc/html/BitSeq.html">BitSeq</a>
transcript expression inference</li>
</ul>

</article>

<article class="" id="slide-13" style="background:;">
  <h3>The generalized linear model</h3>
  <p>\[ K_{ij} \sim \text{NB}( \mu_{ij}, \alpha_i )  \]</p>

<ul>
<li>Read count \(K_{ij}\) for gene <em>i</em> sample <em>j</em>.</li>
<li>2 parameter count distribution: mean \(\mu\), dispersion \(\alpha\)</li>
</ul>

<p>\[ \mu_{ij} = s_{ij} q_{ij} \]</p>

<ul>
<li>Normalized by size factor \(s_{ij}\), with \(q_{ij}\) remaining</li>
<li>Often size factor \(s_j\)</li>
</ul>

</article>

<article class="" id="slide-14" style="background:;">
  <h3>The generalized linear model</h3>
  <p>\[ \log q_{ij} = \sum_r x_{jr} \beta_{ir} \]</p>

<p>\[ \begin{array}{c}
\log q_1 \\
\log q_2 \\
\log q_3 \\
\log q_4 \end{array}
= \left( \begin{array}{cc}
1 & 0 \\
1 & 0 \\
1 & 1 \\
1 & 1 \end{array} \right)
\left( \begin{array}{c}
\beta_0 \\
\beta_1 \end{array} \right) \] </p>

<ul>
<li>results are logarithm base 2 fold changes </li>
</ul>

</article>

<article class="" id="slide-15" style="background:;">
  <h3>Multigroup comparisons in DESeq2</h3>
  <p>\[ \begin{array}{c}
\log q_1 \\
\log q_2 \\
\log q_3 \\
\log q_4 \\
\log q_5 \\
\log q_6 \end{array}
= \left( \begin{array}{cccc}
1 & 1 & 0 & 0 \\
1 & 1 & 0 & 0 \\
1 & 0 & 1 & 0 \\
1 & 0 & 1 & 0 \\
1 & 0 & 0 & 1 \\
1 & 0 & 0 & 1 \\ \end{array} \right)
\left( \begin{array}{c}
\beta_0 \\
\beta_1 \\
\beta_2 \\
\beta_3 \end{array} \right) \] </p>

</article>

<article class="" id="slide-16" style="background:;">
  <h3>Differential expression analysis</h3>
  <pre><code class="r"># takes e.g. ~25 seconds for 8 samples, 35,000 genes
dds &lt;- DESeq( dds )
res &lt;- results( dds )
res &lt;- results( dds, contrast=c(&quot;condition&quot;,&quot;trt&quot;,&quot;untrt&quot;) )
plotMA( res )
</code></pre>

<p><center><img src="plotma.png" width=400/></center></p>

</article>

<article class="" id="slide-17" style="background:;">
  <h3>Plot counts for a single gene</h3>
  <pre><code class="r">gene &lt;- rownames( res )[ order( res$pvalue ) ][ 1 ]
plotCounts( dds, gene, &quot;condition&quot; )
</code></pre>

<p><center><img src="plotcounts.png" width=400/></center></p>

<p>in <em>DESeq2</em> &gt;= v1.5</p>

</article>

<article class="" id="slide-18" style="background:;">
  <h3>Normalization for sample-specific GC and transcript length</h3>
  <ul>
<li><a href="http://www.bioconductor.org/packages/release/bioc/html/cqn.html">cqn</a>
conditional quantile normalization</li>
<li><a href="http://www.bioconductor.org/packages/release/bioc/html/EDASeq.html">EDASeq</a></li>
</ul>

<p><center><img src="cqn.png" width=700/></center></p>

<p><center>cqn package vignette</center></p>

</article>

<article class="" id="slide-19" style="background:;">
  <h3>Controlling for unknown batch</h3>
  <ul>
<li><a href="http://www.bioconductor.org/packages/release/bioc/html/sva.html">sva</a>: <code>svaseq()</code>
surrogate variable analysis</li>
<li><a href="http://www.bioconductor.org/packages/release/bioc/html/RUVSeq.html">RUVSeq</a>:
remove unwanted variation</li>
</ul>

<p>returns a matrix with columns which are surrogate variables</p>

</article>

<article class="" id="slide-20" style="background:;">
  <h3>Controlling for unknown batch</h3>
  <p><center><img src="sva.png"/></center></p>

<p><a href="http://dx.doi.org/10.1371/journal.pgen.0030161">Leek and Storey (2007)</a></p>

</article>

<article class="" id="slide-21" style="background:;">
  <h3>ReportingTools</h3>
  <pre><code class="r">rprt &lt;- HTMLReport(shortName = &quot;analysis&quot;,
                   title = &quot;RNA-Seq analysis&quot;,
                   reportDirectory = &quot;./reports&quot;)
publish(dds, rprt, pvalueCutoff=0.1,
        annotation.db=&quot;org.Hs.eg.db&quot;,
        factor = dds$condition,
        reportDir=&quot;./reports&quot;)
finish(rprt)
</code></pre>

<ul>
<li><a href="http://www.bioconductor.org/packages/release/bioc/html/ReportingTools.html">ReportingTools</a></li>
</ul>

</article>

<article class="" id="slide-22" style="background:;">
  <h3>ReportingTools</h3>
  <p><center><img src="reptools.png" width=700/></center></p>

</article>

<article class="" id="slide-23" style="background:;">
  <h3>Manual annotation</h3>
  <ul>
<li><a href="http://www.bioconductor.org/packages/release/bioc/html/biomaRt.html">biomaRt</a></li>
<li><a href="http://www.bioconductor.org/packages/release/bioc/html/AnnotationDbi.html">AnnotationDbi</a>: <code>select()</code> function, works with the annotation packages <code>org.Hs.eg.db</code>:</li>
</ul>

<pre><code class="r">tab &lt;- select(org.Hs.eg.db, genes, &quot;SYMBOL&quot;, &quot;ENSEMBL&quot;)
</code></pre>

</article>

<article class="" id="slide-24" style="background:;">
  <h3>Support</h3>
  <ul>
<li>mailing list:

<ol>
<li>package name in title</li>
<li>describe experiment, what&#39;s the <em>biological question</em></li>
<li>provide code</li>
<li>sessionInfo()</li>
</ol></li>
<li><code>browseVignettes(&quot;pkg&quot;)</code></li>
<li><code>?function</code></li>
</ul>

</article>

<article class="" id="slide-25" style="background:;">
  <h3>Acknowledgments</h3>
  <ul>
<li>Bioconductor core team</li>
<li><em>DESeq</em>/<em>DEXSeq</em> team

<ul>
<li>Simon Anders</li>
<li>Alejandro Reyes</li>
<li>Wolfgang Huber</li>
</ul></li>
<li>Rafael Irizarry</li>
</ul>

</article>

  </section>
</body>
  <!-- LOAD JAVASCRIPTS  -->
	<script src='libraries/frameworks/html5slides/default/slides.js'></script>
	<!-- MathJax: Fall back to local if CDN offline but local image fonts are not supported (saves >100MB) -->
	<script type="text/x-mathjax-config">
	  MathJax.Hub.Config({
	    tex2jax: {
	      inlineMath: [['$','$'], ['\\(','\\)']],
	      processEscapes: true
	    }
	  });
	</script>
	<script type="text/javascript" src="http://cdn.mathjax.org/mathjax/2.0-latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
	<!-- <script src="https://c328740.ssl.cf1.rackcdn.com/mathjax/2.0-latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
	</script> -->
	<script>window.MathJax || document.write('<script type="text/x-mathjax-config">MathJax.Hub.Config({"HTML-CSS":{imageFont:null}});<\/script><script src="libraries/widgets/mathjax/MathJax.js?config=TeX-AMS-MML_HTMLorMML"><\/script>')
</script>
<!-- LOAD HIGHLIGHTER JS FILES -->
	<script src="libraries/highlighters/highlight.js/highlight.pack.js"></script>
	<script>hljs.initHighlightingOnLoad();</script>
	<!-- DONE LOADING HIGHLIGHTER JS FILES -->
	 
	</html>
